### PR TITLE
Primitive System Remake

### PR DESCRIPTION
**Pull request by ghost**
Remade the primitive system due to issues with the old one, improving performance, documentation, ease of use, and maintainability. Everything using the old system has been changed to use the newer one with no changes to them visually.

The point count on used for DaedalusLightning and DynamicPursuerElectricity has been reduced to a reasonable count as it was excessively high (300 -> 90), and the WulfrumHook's positions used has had Player.Center added to the end to fix it not connecting properly to the player.

### CODE REVIEW

**Review by Ozzatron — CHANGES_REQUESTED** [review]
Please remove the errant localization edits. Everything else looks good. Thank you for your work in improving our archaic primitive rendering system.

**Review by Ozzatron — DISMISSED** [review]

**Review by Ozzatron — DISMISSED** [review]
it appears as though I moved my hand too fast last time

File: Graphics/Primitives/PrimitiveSet.cs [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+		/// The settings of the set.
+		/// </summary>
+		public readonly PrimitiveSettings Settings;
+
+		private PrimitiveSet(VertexPosition2DColorTexture[] vertices, int[] indices, PrimitiveSettings settings)
+		{
+			Vertices = vertices;
+			Indices = indices;
+			Settings = settings;
+		}
```
Comment (steviegt6) [inline Graphics/Primitives/PrimitiveSet.cs]: Fix indentation.

File: Graphics/Primitives/PrimitiveSet.cs [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+        /// <summary>
+        /// Prepares a <see cref="PrimitiveSet"/> based on the provided information, and optionally renders it.
+        /// </summary>
+        /// <param name="positions">The list of positions to use. Keep in mind that these are expected to be in <b>world position</b>, and <see cref="Main.screenPosition"/> is automatically subtracted from them all.</param>
+        /// <param name="settings">The primitive draw settings to use.</param>
+        /// <param name="pointsToCreate">The amount of points to use. More is higher detailed, but less performant. By default, is the number of positions provided. <b>Going above 100 is NOT recommended.</b></param>
+        /// <param name="render">Whether to render the set.</param>
+        /// <returns>The prepared set.</returns>
+        public static PrimitiveSet? Prepare(IEnumerable<Vector2> positions, PrimitiveSettings settings, int? pointsToCreate = null, bool render = true)
+		{
```
Comment (steviegt6) [inline Graphics/Primitives/PrimitiveSet.cs]: tabs

File: Graphics/Primitives/PrimitiveSet.cs [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+            PerformPixelationSafetyChecks(settings);
+
+            int positionCount = positions.Count();
+
+			// Return if not enough to draw anything.
+			if (positionCount <= 2)
+				return null;
```
Comment (steviegt6) [inline Graphics/Primitives/PrimitiveSet.cs]: tabs

File: Graphics/Primitives/PrimitiveSet.cs [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+		}
+        #endregion
+
+        #region Instance Methods
+        /// <summary>
+        /// Renders the provided primitive set to the screen.
+        /// </summary>
+        /// <returns>The provided set.</returns>
+        public PrimitiveSet Render()
+		{
```
Comment (steviegt6) [inline Graphics/Primitives/PrimitiveSet.cs]: tabs

File: Graphics/Primitives/PrimitiveSet.cs [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+            int[] indices = CreateIndicesRectangleTrail(trailPoints.Count);
+
+			// Create the set, and just return it if not asked to render.
+			PrimitiveSet set = new(vertices.ToArray(), indices, settings);
+			if (!render)
+				return set;
+
+			// Else render the set and return it.
+			return set.Render();
+		}
```
Comment (steviegt6) [inline Graphics/Primitives/PrimitiveSet.cs]: tabs

File: Graphics/Primitives/PrimitiveSet.cs [hunk 6]
```diff
@@ -0,0 +1,296 @@
...
+
+            var shaderToUse = Settings.Shader ?? GameShaders.Misc["CalamityMod:StandardPrimitiveShader"];
+            shaderToUse.Shader.Parameters["uWorldViewProjection"].SetValue(view * projection);
+            shaderToUse.Apply();
+
+            // Render the set.
+            Main.instance.GraphicsDevice.DrawUserIndexedPrimitives(PrimitiveType.TriangleList, Vertices, 0, Vertices.Length, Indices, 0, Indices.Length / 3);
+            Main.pixelShader.CurrentTechnique.Passes[0].Apply();
+            return this;
+		}
```
Comment (steviegt6) [inline Graphics/Primitives/PrimitiveSet.cs]: tabs

File: Graphics/Primitives/PrimitiveSet.cs [hunk 7]
```diff
@@ -0,0 +1,296 @@
...
+        #endregion
+
+        #region Set Preperation
+        private static IEnumerable<Vector2> PreparePointsRectangleTrail(IEnumerable<Vector2> positions, PrimitiveSettings settings, int pointsToCreate)
+		{
```
Comment (steviegt6) [inline Graphics/Primitives/PrimitiveSet.cs]: tabs

File: Graphics/Primitives/PrimitiveSet.cs [hunk 8]
```diff
@@ -0,0 +1,296 @@
...
+            // Manually insert the front and end points.
+            points.Insert(0, controlPoints.First());
+            points.Add(controlPoints.Last());
+
+            return points;
+        }
+
+		private static IEnumerable<VertexPosition2DColorTexture> CreateVerticesRectangleTrail(List<Vector2> positions, PrimitiveSettings settings)
+		{
+			List<VertexPosition2DColorTexture> vertices = new();
```
Comment (steviegt6) [inline Graphics/Primitives/PrimitiveSet.cs]: tabs

File: Graphics/Primitives/PrimitiveSet.cs [hunk 9]
```diff
@@ -0,0 +1,296 @@
...
+				indices[startingTriangleIndex] = connectToIndex;
+				indices[startingTriangleIndex + 1] = connectToIndex + 1;
+				indices[startingTriangleIndex + 2] = connectToIndex + 2;
+				indices[startingTriangleIndex + 3] = connectToIndex + 2;
+				indices[startingTriangleIndex + 4] = connectToIndex + 1;
+				indices[startingTriangleIndex + 5] = connectToIndex + 3;
+			}
+
+			return indices;
+		}
```
Comment (steviegt6) [inline Graphics/Primitives/PrimitiveSet.cs]: taabs

File: Graphics/Primitives/IPixelatedPrimitiveRenderer.cs [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+    {
+        /// <summary>
+        /// The layer to render the primitive(s) to.
+        /// </summary>
+        public PixelationPrimitiveLayer LayerToRenderTo => PixelationPrimitiveLayer.BeforeProjectiles;
+
+        /// <summary>
+        /// Render primitives that use pixelation here.
+        /// </summary>
+        public void RenderPixelatedPrimitives(SpriteBatch spriteBatch);
```
Comment (steviegt6) [inline Graphics/Primitives/IPixelatedPrimitiveRenderer.cs]: Do we define access modifiers for other interfaces in Cal? Can we stop doing that if so?

**Review by steviegt6 — CHANGES_REQUESTED** [review]
Mostly just formatting; I don't see anything particularly noteworthy aside from some potential things to take note of if we start facing performance issues (eliminate LINQ, take another look at how the readonly structs are used, they can still be heap-allocated under certain circumstances).

File: Graphics/Primitives/IPixelatedPrimitiveRenderer.cs (hunk 10 above)
Comment (ghost) [reply Graphics/Primitives/IPixelatedPrimitiveRenderer.cs]: Whats the reason they don't have them? I know they can only be public, but it seems weird to me to not include them on interfaces only.

**Review by steviegt6 — APPROVED** [review]

**Review by Ozzatron — APPROVED** [review]

**Comment by HwakYuk** [comment]
Maybe static initialization of `PrimitiveRenderer` needs to be done earlier? The initial call to `PrimitiveRenderer.RenderTrail` in turn invokes the static initializer, but it queues the action to the main thread so that `PrimitiveRenderer.AssignPointsRectangleTrail` results in `NullReferenceException` since `MainPositions` is `null`. This can be reproduced for me by using photon ripper for the first time after starting the game.
It doesn't crash the game, by the way.

**Comment by steviegt6** [comment]
Is it a silently-caught error? We can invoke it at load-time instead, or make it a no-op `ILoadable`.

**Comment by HwakYuk** [comment]
```
[07:17:15.225] [Main Thread/WARN] [tML]: Silently Caught Exception: 
System.NullReferenceException: Object reference not set to an instance of an object.
   at CalamityMod.Graphics.Primitives.PrimitiveRenderer.AssignPointsRectangleTrail(Vector2[] positions, PrimitiveSettings settings, Int32 pointsToCreate) in CalamityMod\Graphics\Primitives\PrimitiveRenderer.cs:line 223
   at CalamityMod.Graphics.Primitives.PrimitiveRenderer.RenderTrail(Vector2[] positions, PrimitiveSettings settings, Nullable`1 pointsToCreate) in CalamityMod\Graphics\Primitives\PrimitiveRenderer.cs:line 97
   at CalamityMod.Projectiles.Melee.PrismTooth.PreDraw(Color& lightColor) in CalamityMod\Projectiles\Melee\PrismTooth.cs:line 135
```
Yes, it is silently caught.

**Pull request merged**